pkg/controller/volume/attachdetach/metrics/metrics.go

```diff
@@ -34,6 +34,19 @@ var (
 		prometheus.BuildFQName("", "storage_count", "attachable_volumes_in_use"),
 		"Measure number of volumes in use",
 		[]string{"node", "volume_plugin"}, nil)
+
+	adControllerVolumesMetric = prometheus.NewGaugeVec(
```

Comment:
A global variable and global methods operating on it is generally not a good idea. It leads to a lot of difficulty unit testing the caller because these dependencies can't be easily replaced. Instead create an interface first defining what methods you need, then create a private struct defining the state you need to implement the interface, then implement a constructor that builds the struct and returns it as an interface parameter, so consumer depends on the interface not the struct implementation.

Reply:
@saad-ali while this is true in other cases - no matter how you slice it, a metric variable is a package global variable and only one instance of it can ever exist in the process. Any attempt to register an already registered metric causes panic errors and process crash. 

From that perspective, i would think that it is better to be explicit and define metrics once in the package and make sure they are initialized/registered only once rather than hide them behind some abstraction, which can potentially cause more errors.  There have been several bugs reported in past which caused controller-manager crash because some code tried to register same metric more than once.